File: python/owrt_monitor/events.py
from __future__ import annotations

import json
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from owrt_monitor.storage import JobStore


class EventLogger:
    def __init__(self, *, store: JobStore, job_id: str, path: Path) -> None:
        self.store = store
        self.job_id = job_id
        self.path = path
        self.path.parent.mkdir(parents=True, exist_ok=True)

    def emit(
        self,
        *,
        level: str,
        component: str,
        event: str,
        message: str,
        fields: dict[str, Any] | None = None,
    ) -> None:
        payload = {
            "ts": datetime.now(UTC).isoformat(),
            "job_id": self.job_id,
            "level": level,
            "component": component,
            "event": event,
            "message": message,
            "fields": fields or {},
        }
        with self.path.open("a", encoding="utf-8") as file:
            file.write(json.dumps(payload, sort_keys=True) + "\n")
        self.store.record_event(
            job_id=self.job_id,
            level=level,
            component=component,
            event=event,
            message=message,
            fields=fields or {},
            ts=payload["ts"],
        )

File: internal/events/events.go
// Package events is the Go port of python/owrt_monitor/events.py. Each Emit
// appends one sorted-key JSON object to events.jsonl AND records the same event
// in the SQLite job_events table, so the daemon's events stream and either
// engine's reader see identical data.
package events

import (
	"encoding/json"
	"os"
	"path/filepath"

	"github.com/tonyandclaw/owrt_monitor/internal/store"
)

// Logger writes structured job events to both events.jsonl and SQLite.
type Logger struct {
	store *store.Store
	jobID string
	path  string
}

// New creates a Logger writing to path (events.jsonl), creating its directory.
func New(st *store.Store, jobID, path string) (*Logger, error) {
	if dir := filepath.Dir(path); dir != "" {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return nil, err
		}
	}
	return &Logger{store: st, jobID: jobID, path: path}, nil
}

// Emit appends one event to events.jsonl and records it in job_events. The
// JSONL object uses sorted keys (Go marshals map keys sorted), matching
// events.py's json.dumps(..., sort_keys=True).
func (l *Logger) Emit(level, component, event, message string, fields map[string]any) error {
	ts := store.NowISO()
	if fields == nil {
		fields = map[string]any{}
	}
	payload := map[string]any{
		"ts":        ts,
		"job_id":    l.jobID,
		"level":     level,
		"component": component,
		"event":     event,
		"message":   message,
		"fields":    fields,
	}
	line, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	f, err := os.OpenFile(l.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	if _, err := f.Write(append(line, '\n')); err != nil {
		_ = f.Close()
		return err
	}
	if err := f.Close(); err != nil {
		return err
	}
	if l.store != nil {
		return l.store.RecordEvent(store.JobEvent{
			JobID: l.jobID, TS: ts, Level: level, Component: component,
			Event: event, Message: message, Fields: fields,
		})
	}
	return nil
}
